### DISCUSSION TITLE
new beta work with browser and nodejs

### DISCUSSION THREAD

**Discussion by lublak** [fabricjs/fabric.js#8870] [Q&A]
Hi. I try to use fabric as a library to work with both browser and nodejs.
But with the new import:
```
import { Canvas, Rect } from 'fabric'; // browser
import { StaticCanvas, Rect } from 'fabric/node'; // node
```
its very hard to get it to work.
I tried to use mostly the imports directly from 'fabric' not 'fabric/node';
The user can pass the StaticCanvas to these. But! It doesn't work. I get:
```
 Uncaught [ReferenceError: window is not defined]
```

only if i use all imports from 'fabric/node' it works fine.
These is not a good solution to work with different environments.
How can i fix this?

**Comment by asturur** [comment]
Sadly node code does not work on browser and vice versa.
A missing dom environment, created through JSDOM on node is what makes the imports differnet.
What is that is making it hard to get it to work? Can you be more specific?

**Reply by lublak** [reply]
> Sadly node code does not work on browser and vice versa.
> A missing dom environment, created through JSDOM on node is what makes the imports differnet.
> What is that is making it hard to get it to work? Can you be more specific?

I try to create a library which renders some stuff on a canvas. So I use StaticCanvas from the fabric package (not node). If I now use these it doesn't work in the node environment.

Yes I can include StaticCanvas from the fabric/node package. But then it doesn't work in the browser. So how can I create a library with:

```ts
import { StaticCanvas} from 'fabric/both`;
function addFancyStuff(canvas: StaticCanvas) {
    canvas.add(...);
}
```

**Reply by ShaMan123** [reply]
With your example simply import it as a type and you are good.
For other cases it is painful, I agree.
Using a mixin is an option as well if you want it to be a class.

**Reply by lublak** [reply]
> With your example simply import it as a type and you are good.
> 
> For other cases it is painful, I agree.
> 
> Using a mixin is an option as well if you want it to be a class.

Sadly no. It's not fine. It's sound crazy but if I use the type from fabric not fabric node it doesn't work. I get this window issue even if I use fabric/node in my main file.

**Reply by ShaMan123** [reply]
```
import type { StaticCanvas } from 'fabric`;
function addFancyStuff(canvas: StaticCanvas) {
    canvas.add(...);
}
```

**Reply by lublak** [reply]
> ```
> import type { StaticCanvas } from 'fabric`;
> function addFancyStuff(canvas: StaticCanvas) {
>     canvas.add(...);
> }
> ```

And here I get a window error. Even if I use it in a nodejs project. Jsdom is installed.
If I change it to:
```ts
import type { StaticCanvas } from 'fabric/node`;
function addFancyStuff(canvas: StaticCanvas) {
    canvas.add(...);
}
```

I can create an example project if I am at home.

**Reply by asturur** [reply]
Canvas is not isomorphic, so it doesn't work. We didn't have it working in fabric 5 either.
What you had to do was installing JSDOM and canvas in node, and carefully leave them out when webpacking your library.
If you wanted both to work you would have to produce 2 different bundles, and let the user choose which one to use.

If you want to have it run in both environment you have to have 2 exports, and one of them has to require the installation of node-canvas and jsdom.

We had a similar issue in fabric 6 and we had to create the `env` import and i think you have to do something similar.

Check what happens in the env folder `src/env/index.ts` and see how it is used in the exports in node.
The difference is that you will need to create a `getFabric` function to use in your code that will return you either fabric from the browser package or the node package, exactly as we did for our env case.

If you find a better solution please tell us.

**Reply by lublak** [reply]
> Canvas is not isomorphic, so it doesn't work. We didn't have it working in fabric 5 either. What you had to do was installing JSDOM and canvas in node, and carefully leave them out when webpacking your library. If you wanted both to work you would have to produce 2 different bundles, and let the user choose which one to use.
> 
> If you want to have it run in both environment you have to have 2 exports, and one of them has to require the installation of node-canvas and jsdom.
> 
> We had a similar issue in fabric 6 and we had to create the `env` import and i think you have to do something similar.
> 
> Check what happens in the env folder `src/env/index.ts` and see how it is used in the exports in node. The difference is that you will need to create a `getFabric` function to use in your code that will return you either fabric from the browser package or the node package, exactly as we did for our env case.
> 
> If you find a better solution please tell us.

Actually, it would be best if Fabric remains independent of the implementation of the canvas api.
```
const canvas = new StaticCanvas(nodeCanvasOrBrowserCanvas);
```

Than you can also use some other canvas based libs:
https://github.com/Brooooooklyn/canvas
https://github.com/joshmarinacci/node-pureimage

**Reply by ShaMan123** [reply]
I like this proposal

**Reply by ShaMan123** [reply]
We discussed it internally.
It isn't only the canvad impl.
There are a few more things, all reside under `src/env`.
So passing the env to the canvas constructor might be a good option.
However, it might be a premature solution because there is a lot of browser specific code that is part of core.
Truth is there is no core concept. That is the real issue. browser, node, etc. should be built on top of that.
